File: twidere/src/main/java/org/mariotaku/twidere/menu/FavoriteItemProvider.java
package org.mariotaku.twidere.menu;

import android.content.Context;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.support.v4.content.ContextCompat;
import android.support.v4.view.ActionProvider;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.widget.ActionMenuView;
import android.view.MenuItem;
import android.view.View;

import org.mariotaku.twidere.graphic.like.LikeAnimationDrawable;
import org.mariotaku.twidere.graphic.like.LikeAnimationDrawable.Style;

import java.lang.ref.WeakReference;

/**
 * Created by mariotaku on 16/2/18.
 */
public class FavoriteItemProvider extends ActionProvider {
    private int mDefaultColor, mActivatedColor;
    private boolean mUseStar;
    private int mIcon;

    /**
     * Creates a new instance.
     *
     * @param context Context for accessing resources.
     */
    public FavoriteItemProvider(Context context) {
        super(context);
    }

    @Override
    public View onCreateActionView() {
        return null;
    }

    public void setUseStar(boolean useStar) {
        mUseStar = useStar;
    }

    public void setDefaultColor(int defaultColor) {
        mDefaultColor = defaultColor;
    }

    public void setActivatedColor(int activatedColor) {
        mActivatedColor = activatedColor;
    }

    public void invokeItem(MenuItem item, LikeAnimationDrawable.OnLikedListener listener) {
        if (MenuItemCompat.getActionProvider(item) != this) throw new IllegalArgumentException();
        final Drawable icon = item.getIcon();
        if (icon instanceof LikeAnimationDrawable) {
            ((LikeAnimationDrawable) icon).setOnLikedListener(listener);
            ((LikeAnimationDrawable) icon).start();
        }
    }

    public void setIcon(int icon) {
        mIcon = icon;
    }

    public void init(final ActionMenuView menuBar, MenuItem item) {
        if (MenuItemCompat.getActionProvider(item) != this) throw new IllegalArgumentException();
        final Drawable icon = ContextCompat.getDrawable(getContext(), mIcon);
        final LikeAnimationDrawable drawable = new LikeAnimationDrawable(icon, mDefaultColor,
                mActivatedColor, mUseStar ? Style.FAVORITE : Style.LIKE);
        drawable.mutate();
        drawable.setCallback(new ViewCallback(menuBar));
        item.setIcon(drawable);
    }

    public void setIsFavorite(MenuItem item, boolean isFavorite) {
        if (MenuItemCompat.getActionProvider(item) != this) throw new IllegalArgumentException();
        final Drawable icon = item.getIcon();
        if (icon instanceof LikeAnimationDrawable) {
            icon.mutate();
            icon.setColorFilter(isFavorite ? mActivatedColor : mDefaultColor, PorterDuff.Mode.SRC_ATOP);
        }
    }

    private static class ViewCallback implements Drawable.Callback {
        private final WeakReference<View> mViewRef;

        public ViewCallback(View view) {
            mViewRef = new WeakReference<>(view);
        }

        @Override
        public void invalidateDrawable(Drawable who) {
            final View view = mViewRef.get();
            if (view == null) return;
            view.invalidate();
        }

        @Override
        public void scheduleDrawable(Drawable who, Runnable what, long when) {
            final View view = mViewRef.get();
            if (view == null) return;
            view.postDelayed(what, when);
        }

        @Override
        public void unscheduleDrawable(Drawable who, Runnable what) {
            final View view = mViewRef.get();
            if (view == null) return;
            view.post(what);
        }
    }
}

File: twidere/src/main/kotlin/org/mariotaku/twidere/menu/FavoriteItemProvider.kt
/*
 *             Twidere - Twitter client for Android
 *
 *  Copyright (C) 2012-2017 Mariotaku Lee <mariotaku.lee@gmail.com>
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.mariotaku.twidere.menu

import android.content.Context
import android.graphics.PorterDuff
import android.graphics.drawable.Drawable
import androidx.core.content.ContextCompat
import androidx.core.view.ActionProvider
import androidx.core.view.MenuItemCompat
import androidx.appcompat.widget.ActionMenuView
import android.view.MenuItem
import android.view.View
import org.mariotaku.ktextension.weak
import org.mariotaku.twidere.BuildConfig
import org.mariotaku.twidere.extension.view.findItemView
import org.mariotaku.twidere.graphic.like.LikeAnimationDrawable
import org.mariotaku.twidere.graphic.like.LikeAnimationDrawable.Style

class FavoriteItemProvider(context: Context) : ActionProvider(context) {
    var defaultColor: Int = 0
    var activatedColor: Int = 0
    var useStar: Boolean = false
    var icon: Int = 0
    var longClickListener: (() -> Boolean)? = null

    override fun onCreateActionView() = null

    fun invokeItem(item: MenuItem, listener: LikeAnimationDrawable.OnLikedListener) {
        if (MenuItemCompat.getActionProvider(item) !== this) throw IllegalArgumentException()
        val icon = item.icon
        if (icon is LikeAnimationDrawable) {
            icon.setOnLikedListener(listener)
            icon.start()
        }
    }

    fun init(menuBar: ActionMenuView, item: MenuItem) {
        if (BuildConfig.DEBUG && MenuItemCompat.getActionProvider(item) !== this) { error("Assertion failed") }
        val icon = ContextCompat.getDrawable(context, this.icon)
        val drawable = LikeAnimationDrawable(icon, defaultColor, activatedColor,
                if (useStar) Style.FAVORITE else Style.LIKE)
        drawable.mutate()
        drawable.callback = ViewCallback(menuBar)
        item.icon = drawable

        val menuView = menuBar.findItemView(item)

        menuView?.setOnLongClickListener { longClickListener?.invoke() == true }
    }

    fun setIsFavorite(item: MenuItem, isFavorite: Boolean) {
        if (MenuItemCompat.getActionProvider(item) !== this) throw IllegalArgumentException()
        val icon = item.icon
        if (icon is LikeAnimationDrawable) {
            icon.mutate()
            icon.setColorFilter(if (isFavorite) activatedColor else defaultColor, PorterDuff.Mode.SRC_ATOP)
        }
    }

    private class ViewCallback(view: View) : Drawable.Callback {
        private val viewRef = view.weak()

        override fun invalidateDrawable(who: Drawable) {
            val view = viewRef.get() ?: return
            view.invalidate()
        }

        override fun scheduleDrawable(who: Drawable, what: Runnable, `when`: Long) {
            val view = viewRef.get() ?: return
            view.postDelayed(what, `when`)
        }

        override fun unscheduleDrawable(who: Drawable, what: Runnable) {
            val view = viewRef.get() ?: return
            view.post(what)
        }
    }
}